### PR TITLE
WM-Actions Send to back functionality

### PR DESCRIPTION
**Pull request by JordanL2**
For feature request #1269 

I also noticed a missing binding removal in fini(), so I added it in a separate commit.

### CODE REVIEW

**Comment by JordanL2** [comment]
Currently has a bug, when it focuses on the top-most view it can focus a view in a different workspace.

**Comment by JordanL2** [comment]
Fixed, although the code is pretty ugly and there might be a better way of checking if two views are on the same workspace.

File: plugins/wm-actions/wm-actions.cpp [hunk 1]
```diff
@@ -267,6 +268,35 @@ class wayfire_wm_actions_t : public wf::plugin_interface_t
...
+            auto views =
+                view->get_output()->workspace->get_views_on_workspace(ws,
+                    wf::LAYER_WORKSPACE);
+            wayfire_view bottom_view = views[views.size() - 1];
+            if (view != bottom_view)
+            {
+                // Move view to bottom by stacking it below the bottom-most view
+                view->get_output()->workspace->restack_below(view, bottom_view);
+
+                // Change focus to the top-most view
```
Comment (ammen99) [inline plugins/wm-actions/wm-actions.cpp:285]: Wayfire's stacking order isn't necessarily indicative of last-focused order. Instead, the next-to-focus view should be the view with the latest `last_focused_timestamp` (see [here](https://github.com/WayfireWM/wayfire/blob/master/src/output/output.cpp#L89)).

**Review by ammen99 — CHANGES_REQUESTED** [review]
LGTM aside from the comment.

File: plugins/wm-actions/wm-actions.cpp (hunk 1 above)
Comment (JordanL2) [reply plugins/wm-actions/wm-actions.cpp]: Done.

**Comment by JordanL2** [comment]
I don't know why the code style check is failing, I've decrusted the only C++ file that's changed here.

**Comment by JordanL2** [comment]
Uh. When I do the code style check command locally:
```
git ls-files | grep "hpp$\|cpp$" | xargs ./uncrustify/build/uncrustify -c uncrustify.ini --check
```
Every file fails except for `plugins/wm-actions/wm-actions.cpp`. I've not changed uncrustify.ini. There hasn't been an uncrustify release since May so it's not a mismatched version.

**Comment by JordanL2** [comment]
Apparently my uncrustify thinks comment lines should be slightly wider. When I try uncrusting a random file this is the change it makes:
```
-/* Returns the intersection of the two boxes, if the boxes don't intersect,
- * the resulting geometry has undefined (x,y) and width == height == 0 */
+/* Returns the intersection of the two boxes, if the boxes don't intersect, the
+ * resulting geometry has undefined (x,y) and width == height == 0 */
```

**Comment by JordanL2** [comment]
I don't know what's wrong, I'm using up-to-date uncrustify from Arch repo and uncrustify.ini is identical to the one on master. I'm uncrusting with:
```
uncrustify -c uncrustify.ini --no-backup plugins/wm-actions/wm-actions.cpp
```
I don't know if it's relevant but this is my LANG env var, in case it's charset related:
```
LANG=en_GB.UTF-8
```

**Comment by JordanL2** [comment]
@ammen99 

Since comment lines are meant to be max 85 chars long:
```
# Try to wrap comments at N columns.
cmt_width                       = 85        # unsigned number
```
My instance of uncrustify actually seems to be working more correctly. Looking at the diff I posted above, the original line is 75 chars wide, the modified line is 79 chars wide. So I think the question is why the CI is limiting comment lines more than needed.

**Comment by JordanL2** [comment]
Fixed my uncrustify issue by compiling uncrustify from source.

File: plugins/wm-actions/wm-actions.cpp [hunk 2]
```diff
@@ -267,6 +268,38 @@ class wayfire_wm_actions_t : public wf::plugin_interface_t
...
+            auto views =
+                view->get_output()->workspace->get_views_on_workspace(ws,
+                    wf::LAYER_WORKSPACE);
+            wayfire_view bottom_view = views[views.size() - 1];
+            if (view != bottom_view)
+            {
+                // Move view to bottom by stacking it below the bottom-most view
+                view->get_output()->workspace->restack_below(view, bottom_view);
+
+                // Change focus to the last focused view on this workspace
```
Comment (ammen99) [inline plugins/wm-actions/wm-actions.cpp:285]: Sorry for the late suggestion, but this is mostly implemented already with this function https://github.com/WayfireWM/wayfire/blob/master/src/api/wayfire/output.hpp#L197

There's a catch though - you should check that there is at least one other view on the workspace, otherwise, `refocus()` will set focus to NULL. But I think that a simple check like that is better than reinventing `refocus()`

Comment (JordanL2) [reply plugins/wm-actions/wm-actions.cpp:285]:
```
  285                 // Change focus to the last focused view on this workspace
  286                 if (views.size() > 1) {
  287                     view->get_output()->refocus(view, wf::LAYER_WORKSPACE);
  288                 }
```
This isn't quite working. If I have three views open: A, B, C (which appear in scale in that order, and I keep triggering the send_to_back, I get:
1. A goes to bottom, B is focused and is the top-most view
2. B goes to bottom, C is the top-most view, but A is focused
Before, I could keep triggering it and I could loop through all three windows in the same order forever, which is the ideal behaviour really. Certainly, it's not acceptable that the top-most view isn't the one focused.

Comment (JordanL2) [reply plugins/wm-actions/wm-actions.cpp:285]: Figured out the problem, don't quite understand it. This change in refocus makes it work:
```
@@ -92,8 +92,9 @@ void wf::output_impl_t::refocus(wayfire_view skip_view, uint32_t layers)
         bool visible_1 = view_on_current_ws(view1);
         bool visible_2 = view_on_current_ws(view2);

-        return std::make_tuple(visible_1, view1->last_focus_timestamp) <
-        std::make_tuple(visible_2, view2->last_focus_timestamp);
+        return (visible_1 && view1->last_focus_timestamp) < (visible_2 && view2->last_focus_timestamp);
```
So, it's the tuple comparison which doesn't seem to be working properly. This "fix" however treats all non-visible views the same, not caring about their last_focus_timestamp.

Comment (JordanL2) [reply plugins/wm-actions/wm-actions.cpp:285]: I've submitted a commit which doesn't use last_focus_timestamp at all and just uses the stacking order as it was before, this is working in my tests.

**Comment by JordanL2** [comment]
I've added another commit, this one emits a `output-stack-order-changed` at the end of the method. This is to alert any relevant plugins that the stack order has changed. Case in point, this is needed for `follow-focus`.

File: plugins/wm-actions/wm-actions.cpp [hunk 3]
```diff
@@ -267,6 +268,36 @@ class wayfire_wm_actions_t : public wf::plugin_interface_t
...
+                views =
+                    view->get_output()->workspace->get_views_on_workspace(ws,
+                        wf::LAYER_WORKSPACE);
+                view->get_output()->focus_view(views[0], false);
+            }
+
+            // Alert any relevant plugins that stack order has changed
```
Comment (ammen99) [inline plugins/wm-actions/wm-actions.cpp:292]: workspace-manager should emit this signal when restacking the view below, right? Why is this needed again here?

File: plugins/wm-actions/wm-actions.cpp (hunk 2 above)
Comment (ammen99) [reply plugins/wm-actions/wm-actions.cpp:285]: Okay, I guess it makes sense that we don't follow last-focused order when we're changing the stacking order. I'm leaving this for you to decide, current version LGTM.

File: plugins/wm-actions/wm-actions.cpp (hunk 3 above)
Comment (JordanL2) [reply plugins/wm-actions/wm-actions.cpp:292]: This was to make it work better with follow-focus so the view the cursor is pointing to gets the focus. We have to do this after the focus command of course. However, I'm not sure if we actually want to do the follow-focus change I submitted any more, after thinking about it more. So we possibly should take this signal out for now.

Comment (JordanL2) [reply plugins/wm-actions/wm-actions.cpp]: Removed the above change.

**Review by ammen99 — APPROVED** [review]
Thanks for the patience, LGTM now!
